File: solutions/python/code-carl/ch09_dp/t11_lc0494_find_target_sum_ways.py
"""LeetCode 494. Target Sum
https://leetcode.cn/problems/target-sum/
"""
from symbol import yield_arg
from typing import List


class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums)
        if abs(target) > total or (total + target) % 2 == 1:
            return 0

        bag_size = (target + total) // 2
        dp: List[int] = [0] * (bag_size + 1)
        dp[0] = 1

        for n in nums:
            for j in range(bag_size, n - 1, -1):
                dp[j] += dp[j - n]

        return dp[-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.findTargetSumWays([1, 1, 1, 1, 1], 3) == 5
    assert solution.findTargetSumWays([1], 1) == 1
    print("All tests passed.")

File: solutions/code-carl/ch09_dp/T11_LC0494_FindTargetSumWays.java
/**
 * Practice template for:
 * - LeetCode 494. Target Sum
 *   https://leetcode.cn/problems/target-sum/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/22 15:36
 * @comment
 */
public class T11_LC0494_FindTargetSumWays {
    public int findTargetSumWays(int[] nums, int target) {
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }
        if ((target + sum) % 2 == 1 || Math.abs(target) > sum) {
            return 0;
        }
        int bagSize = (target + sum) / 2;

        int[] dp = new int[bagSize + 1];
        dp[0] = 1;
        for (int i = 0; i < nums.length; i++) {
            for (int j = bagSize; j >= nums[i]; j--) {
                dp[j] += dp[j - nums[i]];
            }
        }
        return dp[bagSize];
    }
}
